### DISCUSSION TITLE
MCP230xx Hardware Interrupts

### DISCUSSION THREAD

**Discussion by jinjamator** [arendst/Tasmota#14024] [Q&A] [answered]
Is there a reason not using the hardware interrupt lines of the MCP230xx chips, or is it just not implemented?

**Comment by plewka** [comment, marked as answer]
In my understanding it's up to you to connect the IRQ line to an input and handle it?
This is a generic question which is not dependent on the MCP230xx, isn't it?

In I2C the IRQ-pin (if available) is a side-channel to inform the microprocessor 
something defined happened in peripheral. Often the IRQs are tied together to 
create a wired-OR for the CPU done to prevent unnecessary polling or to shorten 
latency.

**Reply by jinjamator** [reply]
Ok I refine my question. :) The Tasmota included MCP230XX I2C driver does not support the IRQ Lines of the chips. My question is if this has a specific reason or was it just not implemented as polling the interrupt register of the chip every 50ms is good enough.
I' thinking about adding support for this, but if it was implemented because it makes no sense because of reasons, I'll use my time for something else.

**Reply by barbudor** [reply]
Tasmota is not a hard real-time system. It's based on a polling loop and some kind of message queue.
If you were to get interrupts for MCP pins change, at least you would still have to post to the message queue which could be complicated as the code as not been designed to be re-entrant (because not necessary).
As far as I can see, there would be very little use-cases where interrupts could be usefull.
But I may be wrong, so if you have such use case, please elaborate.

**Reply by jinjamator** [reply]
I.C. I don't really have a use case. I was just thinking interrupts are more efficient why don't use them when they are there.

**Reply by sfromis** [reply]
Efficiency hardly ever matters for simple on/off pin control on a dedicated MCU, but the complexity of handling interrupts is arguably not at all efficient, when programming time is also factored in. For Tasmota, the major efficiency factor is not saving processor time, but cutting back on bytes used in RAM and code.

**Reply by jinjamator** [reply]
I meant efficent in cpu cycles and power usage. But as I can see the focus of tasmota drives are other resources, which is totally fine for me.
